### PR TITLE
Allow `obsm` and `varm` to be stored sparsely

### PR DESCRIPTION
**Pull request by johnkerl**
As discussed in the March 31 2023 spec meeting.

See also https://github.com/single-cell-data/TileDB-SOMA/issues/1245.

### CODE REVIEW

**Comment by johnkerl** [comment]
@atolopko-czi @mlin @maniarathi ping 🙏

**Review by aaronwolen — APPROVED** [review]

File: abstract_specification.md [hunk 1]
```diff
@@ -270,9 +270,9 @@ The `SOMAMeasurement` is a sub-element of a `SOMAExperiment`, and is otherwise a
...
 | `X`        | `SOMACollection[string, SOMASparseNDArray\|SOMADenseNDArray]` | A collection of matrices, each containing measured feature values. Each matrix is indexed by `[obsid, varid]`. Both sparse and dense 2D arrays are supported in `X`.                                                                                                                               |
-| `obsm`     | `SOMACollection[string, SOMADenseNDArray]`                    | A collection of dense matrices containing annotations of each _obs_ row. Has the same shape as `obs`, and is indexed with `obsid`.                                                                                                                                                                 |
+| `obsm`     | `SOMACollection[string, SOMASparse\|SOMADenseNDArray]`        | A collection of dense matrices containing annotations of each _obs_ row. Has the same shape as `obs`, and is indexed with `obsid`.                                                                                                                                                                 |
```
Comment (thetorpedodog) [inline abstract_specification.md:273]: should probably say `SOMASparseNDArray|SOMADenseNDArray`

**Review by thetorpedodog — APPROVED** [review]

**Review by atolopko-czi — CHANGES_REQUESTED** [review]
The changes LGTM, but do we also need to make other SOMA-repo related changes in the same issue/PR? I'm seeing these, for one:
*. https://github.com/single-cell-data/SOMA/blob/b1c7463de1dffef504b54f08dda714c1bafcc814/python-spec/src/somacore/measurement.py#L79
* https://github.com/single-cell-data/SOMA/blob/b1c7463de1dffef504b54f08dda714c1bafcc814/python-spec/src/somacore/measurement.py#L91
If there is another issue to make code-related changes, I will approve.

**Comment by thetorpedodog** [comment]
> The changes LGTM, but do we also need to make other SOMA-repo related changes in the same issue/PR?

These, fortunately, are trivial (just subbing in a different type; no actual runtime behavior changes) and IMO fine for a followup.

**Review by atolopko-czi — APPROVED** [review]

File: abstract_specification.md (hunk 1 above)
Comment (johnkerl) [reply abstract_specification.md:273]: @thetorpedodog no. the change is to allow dense `obsm`/`varm` to be sparse. not the other way around for `obsp`/`varp`.

Comment (johnkerl) [reply abstract_specification.md:273]: @thetorpedodog outdated comment? I can't find this

**Pull request merged**